src/metabase/query_processor/preprocess.clj

```diff
@@ -39,56 +41,75 @@
    [metabase.query-processor.middleware.validate-temporal-bucketing :as validate-temporal-bucketing]
    [metabase.query-processor.middleware.wrap-value-literals :as qp.wrap-value-literals]
    [metabase.query-processor.setup :as qp.setup]
+   [metabase.query-processor.store :as qp.store]
    [metabase.util :as u]
    [metabase.util.i18n :as i18n]
    [metabase.util.log :as log]
    [metabase.util.malli :as mu]))
 
+;;; the following helper functions are temporary, to aid in the transition from a legacy MBQL QP to a pMBQL QP. Each
+;;; individual middleware function is wrapped in either [[ensure-legacy]] or [[ensure-pmbql]], and will then see the
+;;; flavor of MBQL it is written for.
+
+(mu/defn ^:private ->legacy :- mbql.s/Query
+  [query]
+  (lib.convert/->legacy-MBQL query))
+
+(defn- ^:deprecated ensure-legacy [middleware-fn]
+  (fn [query]
+    (let [query (cond-> query
+                  (:lib/type query) ->legacy)]
+      (middleware-fn query))))
+
+(defn- ensure-pmbql [middleware-fn]
+  (fn [query]
+    (let [query (cond->> query
+                  (not (:lib/type query)) (lib.query/query (qp.store/metadata-provider)))]
+      (middleware-fn query))))
+
 (def ^:private middleware
   "Pre-processing middleware. Has the form
 
     (f query) -> query"
   ;; ↓↓↓ PRE-PROCESSING ↓↓↓ happens from TOP TO BOTTOM
+  #_{:clj-kondo/ignore [:deprecated-var]}
   [#'normalize/normalize-preprocessing-middleware
-   #'qp.perms/remove-permissions-key
-   #'qp.constraints/maybe-add-default-userland-constraints
-   #'validate/validate-query
-   #'fetch-source-query/resolve-source-cards
-   ;; ↑↑↑ ALL MIDDLEWARE ABOVE THIS POINT WILL SEE MLV2 PMBQL QUERIES ↑↑↑
-   #'qp.convert-to-legacy/convert-to-legacy
-   ;; ↓↓↓ ALL MIDDLEWARE BELOW THIS POINT WILL SEE LEGACY MBQL QUERIES ↓↓↓
-   #'expand-macros/expand-macros
-   #'qp.resolve-referenced/resolve-referenced-card-resources
-   #'parameters/substitute-parameters
-   #'qp.resolve-source-table/resolve-source-tables
-   #'qp.auto-bucket-datetimes/auto-bucket-datetimes
-   #'reconcile-bucketing/reconcile-breakout-and-order-by-bucketing
-   #'qp.add-source-metadata/add-source-metadata-for-source-queries
-   #'upgrade-field-literals/upgrade-field-literals
-   #'qp.middleware.enterprise/apply-sandboxing
-   #'qp.persistence/substitute-persisted-query
-   #'qp.add-implicit-clauses/add-implicit-clauses
-   #'qp.add-dimension-projections/add-remapped-columns
-   #'qp.resolve-fields/resolve-fields
-   #'binning/update-binning-strategy
-   #'desugar/desugar
-   #'qp.add-default-temporal-unit/add-default-temporal-unit
-   #'qp.add-implicit-joins/add-implicit-joins
-   #'resolve-joins/resolve-joins
-   #'resolve-joined-fields/resolve-joined-fields
-   #'fix-bad-refs/fix-bad-references
-   #'escape-join-aliases/escape-join-aliases
+   (ensure-pmbql #'qp.perms/remove-permissions-key)
```

Comment:
I changed this to be wrappers around all the middleware so we can port the middleware in whatever order we feel like instead of being forced to do it in a certain order. This will make things easier going forward since some middleware uses functions from other middleware